Py3 only fixes breaking Py2 compatibility

mliu49 opened the pull request:
This PR contains changes necessary for full compatibility with Python 3, some of which require breaking Python 2 compatibility.

#1683, which replaces `dde` with `chemprop`, has been included in this PR. The purpose is to have a single PR which switches (the `py3` branch) from being fully functional in Python 2 to fully functional in Python3.

As far as testing goes, I think the requirement for merging this into the `py3` branch will simply be reviewer approval and passing standard Travis tests. Once that's done, more rigorous testing will be done on the `py3` branch.

mliu49 commented:
### Unresolved Issue 1: XYZ string generation in Arkane
```
======================================================================
FAIL: Test properly loading the ArkaneSpecies object and respective sub-objects
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mjliu/Code/RMG-Py/arkane/commonTest.py", line 369, in test_create_and_load_yaml
    self.assertTrue('C 7.54e-14 1.193e-13 5.52e-14' in arkane_spc.xyz)
AssertionError: False is not true

----------------------------------------------------------------------
```
I commented out the failing assertion temporarily so that unit tests will pass and RMG-tests will run. The cause has been identified as a change in how floats are handled by `str.format`.

### Unresolved Issue 2: Sorting Entry objects
I just discovered this while running a test job. I think we might need to do a global search for sorting and verify that all instances of implicit sorting of custom classes are addressed.
```
Traceback (most recent call last):
  File "/home/mjliu/Code/RMG-Py/rmg.py", line 185, in <module>
    main()
  File "/home/mjliu/Code/RMG-Py/rmg.py", line 179, in main
    rmg.execute(**kwargs)
  File "/home/mjliu/Code/RMG-Py/rmgpy/rmg/main.py", line 894, in execute
    trimolecularReact=self.trimolecularReact)
  File "/home/mjliu/Code/RMG-Py/rmgpy/rmg/model.py", line 646, in enlarge
    self.updateUnimolecularReactionNetworks()
  File "/home/mjliu/Code/RMG-Py/rmgpy/rmg/model.py", line 1756, in updateUnimolecularReactionNetworks
    network.update(self, self.pressureDependence)
  File "/home/mjliu/Code/RMG-Py/rmgpy/rmg/pdep.py", line 776, in update
    spec.generateStatMech()
  File "rmgpy/species.py", line 779, in rmgpy.species.Species.generateStatMech
  File "/home/mjliu/Code/RMG-Py/rmgpy/data/statmech.py", line 667, in getStatmechData
    statmech_model = self.getStatmechDataFromGroups(molecule, thermoModel)
  File "/home/mjliu/Code/RMG-Py/rmgpy/data/statmech.py", line 701, in getStatmechDataFromGroups
    return self.groups['groups'].getStatmechData(molecule, thermoModel)
  File "/home/mjliu/Code/RMG-Py/rmgpy/data/statmech.py", line 404, in getStatmechData
    for entry in group_count if group_count[entry] > 0])
TypeError: '<' not supported between instances of 'Entry' and 'Entry'
```

alongd commented:
Why is the coverage significantly lower?

mliu49 commented:
Good question! It might be due to the fact that codecov is comparing between py2 results and py3 results, although it's not apparent what differences would affect coverage that much.

mliu49 commented:
Rebased onto `py3`. The two unresolved issues mentioned earlier have been fixed.

amarkpayne commented on rmgpy/data/statmech.py:
```diff
@@ -400,8 +400,9 @@ def getStatmechData(self, molecule, thermoModel):
                 while freq_count > num_vibrations:
-                    min_degeneracy, min_entry = min([(entry.data.symmetry, entry)
-                                                     for entry in group_count if group_count[entry] > 0])
+                    min_entry = min((entry for entry in group_count if group_count[entry] > 0),
```
Would it be better to implement the comparison methods for the class or is it more likely that the sorting keys will be case-by-case dependent

mliu49 replied:
It seemed like Entries would generally be sorted for a specific purpose, which would require a custom sorting key. Species/Molecules/Atoms/Bonds were often sorted for deterministic-ness, so it seemed appropriate to implement comparison methods.

cgrambow commented on rmgpy/ml/estimator.py:
```diff
@@ -28,17 +25,24 @@
 #                                                                             #
 ###############################################################################
 
+import contextlib
```
They are organized. See [this discussion](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/1683#discussion_r314728958).

amarkpayne replied:
You are correct. This was either outdated or (more likely) I was just mistaken. Sorry about that

cgrambow commented on rmgpy/ml/estimator.py:
```diff
@@ -48,55 +52,41 @@ class MLEstimator():
         cp0 = molecule.calculateCp0()
         cpinf = molecule.calculateCpInf()
+
+        # Set uncertainties to 0 because the current model cannot estimate them
```
Great question. Gabriele and Adam were/are working on uncertainty estimation, which is actually really difficult to do correctly, but I'm confident that something suitable for RMG will be available in the near future.

mliu49 commented on rmgpy/data/test_data/thermo/groups/adsorptionPt.py:
```diff
@@ -2,8 +2,8 @@
 # encoding: utf-8
 
 name = "Surface Adsorption Corrections"
-shortDesc = u""
```
We should remove them from RMG-database, but it's not crucial. I removed them from the testing databases because it was straightforward to do.

Ideally, we would machine write the entire database, like we planned to do a while back. Another consideration is whether or not we want to change the database syntax.

mliu49 commented on .travis.yml:
```diff
@@ -39,38 +32,36 @@ before_install:
       install:
         - cd RMG-Py
-        - conda env create -q -f environment_linux.yml
+        - conda env create -q -f environment_py3.yml
```
It could be part of this PR. I started working on it, but then realized that tests and documentation would need to be updated as well.

My plan is to remove the old environment files completely and rename `environment_py3.yml` to `environment.yml`.

mliu49 commented on rmgpy/molecule/molecule.py:
```diff
@@ -551,6 +569,34 @@ def __reduce__(self):
+        Define a custom hash method to allow Bond objects to be used in dictionaries and sets.
+        """
+        return hash(('Bond', self.order,
+                     self.atom1.symbol if self.atom1 is not None else '',
```
I'm not sure, to be honest. It seems like it would make sense to have C-H and H-C return the same hash, but right now equality is defined as a memory id comparison, so this hash would provide acceptable behavior with fewer collisions. I'm open to changing it though.

amarkpayne replied:
Let's keep it as is then, since it results in less collisions and we have memory id comparisons for equality checks.

mliu49 commented on rmgpy/rmg/model.py:
```diff
@@ -1928,9 +1928,12 @@ def generateReactionKey(rxn, useProducts=False):
+    if len(spc_list) == 1:
+        key1, key2 = getKey(spc_list[0]), None
+    else:
+        # Use 2 keys, even for trimolecular reactions?
```
We definitely supported trimolecular library reactions at the time this was written, so it might have been considered. I feel like it would be more efficient to have 3 keys for trimolecular. With that said, I think it's probably not a good idea to change this now, and instead revisit it after py3.

mliu49 commented on setup.py:
```diff
@@ -218,7 +218,7 @@
 ]
 
-modules = []
+modules = ['rmg.py']
```
Did you test that this works? I think it should just be `rmg`, but I'm also not sure if you can specify it as both a module and a script.

amarkpayne replied:
Just to document, we scraped this solution and instead have move the function in rmg.py to util.py where it can be imported.

mliu49 commented:
I got this warning while running a test job on this branch:
```
/home/mjliu/Code/RMG-Py/rmgpy/tools/plot.py:211: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure()
```

Maybe this resulted from switching some plotting functionality from pylab to pyplot? Regardless, we should probably explicitly close the figure anytime we plot something, using `matplotlib.pyplot.close`.

mliu49 commented on Makefile:
```diff
@@ -50,6 +50,9 @@ endif
 
 q2dtor:
+ifneq ($(wildcard external/Q2DTor/.),)
+	@ echo "Q2DTor found in external/Q2DTor. Please make sure that the branch 'arkanepy3' is checked out for this."
```
Thinking about it more, this doesn't make a lot of sense. If a user already has q2dtor, they won't have any reason to run `make q2dtor` again.

I think we can either
a) do nothing, assume that most users have not installed it yet, and tell people how to fix it if they ask
b) suggest double-checking the branch if running q2dtor fails by catching the error

amarkpayne replied:
I guess I was thinking that the user might be tempted to run make q2dtor to fix issues related to q2dtor in the same way that `make clean; make` fixes a lot of issues. I was going to add code here that checked out the branch automatically, but I guess I forgot to include it.

I'll go with suggestion a) then

mliu49 commented:
I went ahead and squashed the fixes and rebased. I would like to get this in asap so we can move on to finalizing other stuff.

One last thing to address is how we want to handle the two temp commits. Since this branch incorporates chemprop and makes changes to environment files, both RMG-database and RMG-tests require updates.

My proposal is to add the chemprop models into RMG-database without removing the dde models yet, so that python 2 and python 3 builds will work, and to modify RMG-tests to use `environment.yml` when available and `environment_[os].yml` otherwise. This way, all tests will pass. We can then remove the dde models from database at a later date.

@cgrambow, any thoughts about database?

amarkpayne commented on rmgpy/ml/estimator.py:
```diff
@@ -1,6 +1,3 @@
-#!/usr/bin/env python
```
I might have things mixed up, but isn't there a difference between unicode and UTF-8? I thought that unicode was just a matching of characters to unique indexes (codepoints) and that UTF-8 was one of a few different ways of encoding unicode characters (for example there is also UTF-16 and others). In fact, it appears that python behind the scenes is not using UTF-8 for it's encoding because it needs every character to take up the same number of bytes to be able to do random access to individual characters in strings. (see https://rushter.com/blog/python-strings-and-memory/).

That being said, UTF-8 is almost ubiquitous, so it is unlikely that someone would have trouble with this (I think most OSs use UTF-8 by default and if not most things are good about being able to automatically figure out encodings when they are standard like UTF-8), but I don't think we need to get rid of these statements either.

My interpretation of this could be wrong though, so please correct me if so.

amarkpayne replied:
Does this just use the system default encoding? Or does python set this as the default regardless of the OS? Either way I am almost certain that UTF-8 is the default encoding for mac and Linux, so we are fine.

Again, I don't think this is an issue, just curious to know.

cgrambow replied:
I think you're mixing up things. The UTF-8 magic comment simply tells the Python interpreter to assume the file is in UTF-8, which it now does by default in Python 3. That doesn't actually mean the file is UTF-8 which depends on how the editor you are using encodes the file, so it's up to the programmer to ensure that their editor is set up correctly.

cgrambow replied:
As @amarkpayne and I just discussed, it could still be useful to keep the comment in if editors rely on reading it to determine encoding. However, PyCharm, vim, emacs, probably many others don't need it.

cgrambow commented:
I'd say just kick out the DDE models but no strong feelings either way.

amarkpayne commented:
I made a quick comment about the file encoding lines, but otherwise once the TEMP commit issues get resolved I approve of merging this PR into py3 as long as @alongd and @cgrambow feel the same way.

I'll work with @mliu49 to review and merge the needed changes for RMG-tests

mliu49 commented:
If we remove the dde models, all master based RMG-Py builds, RMG-database builds, and RMG-tests builds will fail until we merge the py3 branch. If anyone were to pull current database during this time, their install would also be broken. If the time between merging this PR into py3 and merging py3 into master were to be less than 24 hours, I think it would be acceptable, but I feel like that is unlikely.
